edb/edgeql/parser/grammar/expressions.py

```diff
@@ -549,135 +551,86 @@ class PtrQualsSpec(typing.NamedTuple):
     cardinality: typing.Optional[qltypes.SchemaCardinality] = None
 
 
-class PtrQuals(Nonterm):
-    def reduce_OPTIONAL(self, *kids):
-        self.val = PtrQualsSpec(required=False)
+class DummyEmpty(Nonterm):
+    def reduce_empty(self, *kids):
+        self.val = None
 
-    def reduce_REQUIRED(self, *kids):
-        self.val = PtrQualsSpec(required=True)
 
-    def reduce_SINGLE(self, *kids):
-        self.val = PtrQualsSpec(cardinality=qltypes.SchemaCardinality.One)
+# Class decorator that inlines any references to InlinedPtrQuals in
+# a production.
+def inline_ptr_quals(cls):
+    required = [('OPTIONAL', False), ('REQUIRED', True), ('<e>', None)]
+    cards = [
+        ('SINGLE', qltypes.SchemaCardinality.One),
+        ('MULTI', qltypes.SchemaCardinality.Many),
+        ('<e>', None),
+    ]
 
-    def reduce_MULTI(self, *kids):
-        self.val = PtrQualsSpec(cardinality=qltypes.SchemaCardinality.Many)
+    for name, func in list(cls.__dict__.items()):
+        if not (
+            name.startswith('reduce_') and isinstance(func, types.FunctionType)
+        ):
```

Comment:
I really don't think this is a good idea.

It is more concise, but it is also magic that takes a few minutes to understand.